File: Medium/1609-Even Odd Tree/Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: TreeNode) -> bool:
        even_index = True
        curr_level = [root]

        while curr_level != []:
            prev_val = 0 if even_index else 1_000_001
            next_level = []

            for node in curr_level:
                if (even_index and
                    (node.val % 2 == 0 or prev_val >= node.val)) or \
                        (not even_index and
                         (node.val % 2 == 1 or prev_val <= node.val)):
                    return False

                prev_val = node.val
                if node.left is not None:
                    next_level.append(node.left)
                if node.right is not None:
                    next_level.append(node.right)

            even_index = not even_index
            curr_level = next_level

        return True

File: Medium/1609-Even Odd Tree/Level Order Traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_even_odd_tree(root)
  even_index = true
  curr_level = [root]

  until curr_level.empty?
    prev_val = even_index ? 0 : 1_000_001
    next_level = []

    curr_level.each do |node|
      return false if even_index && (node.val.even? || prev_val >= node.val)
      return false if !even_index && (node.val.odd? || prev_val <= node.val)

      prev_val = node.val
      next_level.push(node.left) unless node.left.nil?
      next_level.push(node.right) unless node.right.nil?
    end

    even_index = !even_index
    curr_level = next_level
  end

  true
end
